### PR TITLE
Reject other QASM versions in QASM 2 parser

### PR DESCRIPTION
**Pull request by 3yakuya**
### Summary
Fixing [QASM-319](https://github.com/Qiskit/openqasm/issues/319), qasm parser will fail with appropriate error if invalid OPENQASM version is specified. Reno release note is added.

### Details and comments

Small change in qasm lexer to only accept 2.0 as a version, rather than any digit.digit. The latter caused a bug where user could provide any version, like `3.0` and believe it works while in reality their code was still parsed as `2.0`.

Should not break any workflows, unless they rely on broken openqasm strings. Added reno release note. Did not see any place in the documentation per se that would require updating.

### CODE REVIEW

**Comment by 3yakuya** [comment]
Its effectively blocked by https://github.com/Qiskit/qiskit-terra/issues/7457

File: qiskit/qasm/qasmlexer.py
Comment (jakelishman) [inline qiskit/qasm/qasmlexer.py:164]: I think it would be good if we update the error messages such that if you try and parse a programme beginning with `OPENQASM 3.0;` or similar, the error tells you that this module is for QASM 2 only, and add a test that that works correctly.  This might require keeping the token the way it currently is, and handling it later on, during parsing.  At any rate, it'd be good to clean up `QasmParser.p_format_0`, either by absorbing it into `p_format` to actually perform the check, or removing it if we're not going to use it at all.

File: releasenotes/notes/qasm-lexer-bugfix-1779525b3738902c.yaml
Comment (jakelishman) [inline releasenotes/notes/qasm-lexer-bugfix-1779525b3738902c.yaml:3]: We should be clear that only the QASM 2 tools will error when passed versions other than 2.0; the export capabilities for QASM 3 live in `qiskit.qasm3` - a separate module.

File: releasenotes/notes/qasm-lexer-bugfix-1779525b3738902c.yaml
Comment (jakelishman) [inline releasenotes/notes/qasm-lexer-bugfix-1779525b3738902c.yaml:4]: I've transferred this issue into Terra, so the issue link is out-of-date, but in this case the buggy behaviour is sufficiently simple, we'd be ok without a link.

**Review by jakelishman — COMMENTED** [review]
Thanks for looking at this!

If we're making more programmes fail, we should make sure that our error messages are particularly clear - in this case, it'd be good if we can ensure that the error tells us that only parsing of QASM 2 is currently supported, if we encounter a bad `OPENQASM` statement.  At the moment, the error is:
```
QasmError: 'Unable to match any token rule, got -->O<-- Check your OPENQASM source and any include statements.'
```
which isn't super helpful.

**Comment by 3yakuya** [comment]
Makes sense, thanks! Will try to look late today or early tomorrow

**Comment by 3yakuya** [comment]
@jakelishman any chance for some quick eyes here?

File: qiskit/qasm/qasmparser.py
Comment (jakelishman) [inline qiskit/qasm/qasmparser.py:379]: This new form works rather well, thanks!  Can we put in the parsed version token (e.g. "Invalid version string: ... This module only supports OpenQASM 2") too?  Also, just to be a little more permissive - while OpenQASM 2 techncially only allows the version to be `2.0`, OpenQASM 3 currently allows it to be a floating point number (though in practice we ought to change that to something more suited to representing versions).  That means that `OPENQASM 3;` should be a valid header for OpenQASM 3 programs, but it'll throw the old
```python
QasmError: 'Unable to match any token rule, got -->O<-- Check your OPENQASM source and any include statements.'
```
error here.  We could relax the parser rule to `OPENQASM\s+[0-9]\+(\.[0-9]\+)?'` here, and default to `minor_version = 0` if we don't match that group.  (We'll need tests for `OPENQASM 3;` throwing, and `OPENQASM 2;` being accepted)

I think that's as far as I'd want to go - in theory it might be possible to break the parsing with a super weird error message if we put in a _token_ that OpenQASM 3 has, but 2 doesn't (like funny Unicode stuff), but off a very brief attempt, that seemed to work fine for me here.

File: releasenotes/notes/qasm-lexer-bugfix-1779525b3738902c.yaml
Comment (jakelishman) [inline releasenotes/notes/qasm-lexer-bugfix-1779525b3738902c.yaml:6]:
```suggestion
    Stopped the parser in :meth:`.QuantumCircuit.from_qasm_str` and
    :meth:`~.QuantumCircuit.from_qasm_file` from accepting OpenQASM programs
    that identified themselves as being from a language version other than 2.0.
    This parser is only for OpenQASM 2.0; support for imported circuits from
    OpenQASM 3.0 will be added in an upcoming release.
```

**Review by jakelishman — COMMENTED** [review]
Sorry this took so long for me to look at again.  The errors look much better now, in most of the situations, I just had one little tweak for a new situation that OpenQASM 3 allows, but OpenQASM 2 doesn't, that'll improve them a little more too.

We're good to go on this after that.

File: qiskit/qasm/qasmparser.py
Comment (3yakuya) [reply qiskit/qasm/qasmparser.py:379]: Got it, makes sense. Pushing this now.

File: qiskit/qasm/node/format.py
Comment (jakelishman) [inline qiskit/qasm/node/format.py:32]: You probably meant to match group 4 here (group 3 will include the dot).  You can also avoid the extra boolean value:
```python
self.minorversion = parts.group(4) if parts.group(4) is not None else "0"
```
would be fine as `Match.group` returns `None` for a valid group that isn't matched.  Your current `has_minor_version` is actually wrong; `re.match` searches only for matches at the start of the string, and since it's only valid to be here if the string starts with `OPENQASM`, `has_minor_version` can't be true.

(Entirely up to you, but if you don't like the group numbers, you can also name groups in `re` objects.  You can do 
```python
parts = re.match(r"OPENQASM\s+(?P<major>\d+)(\.(?P<minor>\d+))?", value)
```
and then it's valid to access `parts.group("major")` and `parts.group("minor")`).

File: test/python/test_qasm_parser.py
Comment (jakelishman) [inline test/python/test_qasm_parser.py:89]: I know you were copying the other functions in this module (which is a good thing in general), but there's a clearer way of writing these that I'm slowly trying to transition the tests to when appropriate:
```python
with self.assertRaisesRegex(QasmError, "Invalid version string <...>"):
    parse(filename)
```

Personally, I'd not wrap this up in a separate function - it's very short code used only twice, and in tests it's not a bad thing to be a little verbose in order to see exactly what's being tested.  The way the tests are parameterised in this class is rather lacking (not your fault at all, but we absolutely should not need separate files here, and calculating the directory only during instance setup is a nuisance), but we could remove the need for this function by doing
```python
import ddt

@ddt.ddt
class TestParser:
    ...

    @ddt.data("example_version_fail.qasm", "example_minor_version_fail.qasm")
    def test_parser_version_fail(self, filename):
        filename = os.path.join(self.qasm_dir, filename)
        with self.assertRaisesRegex(QasmError, "Invalid version string <...>"):
            parse(filename)
```
Now it's clear that I'm running multiple tests of the same thing, and the parametrisation is explicitly at listed at the call site.

File: test/python/qasm/example_version_2.qasm
Comment (jakelishman) [inline test/python/qasm/example_version_2.qasm:1]: The fact that `OPENQASM 2;` works now is kind of a side-effect - it'd be best for the good test file to use `2.0`.

File: test/python/qasm/example_version_fail.qasm
Comment (jakelishman) [inline test/python/qasm/example_version_fail.qasm:1]: You already test `OPENQASM 3;` so this isn't testing anything new; a good other test would be that minor versions of `2` are also detected, such as `OPENQASM 2.1;` - this would have caught the error with `has_minor_version` above, because the file would have parsed successfully when it should have thrown an error.

File: test/python/qasm/example_version_2.qasm
Comment (3yakuya) [reply test/python/qasm/example_version_2.qasm:1]: For this one we have a few tests that run with 2.0 and I wanted to keep at least one that shows that indeed a single-digit version will work

File: test/python/qasm/example_version_fail.qasm
Comment (3yakuya) [reply test/python/qasm/example_version_fail.qasm:1]: Eh you got more work than help with my contribution here, sorry for rookie mistakes! Working on it

File: qiskit/qasm/node/format.py
Comment (3yakuya) [reply qiskit/qasm/node/format.py:32]: 🤦🏻‍♂️  of course, good catch and thank you.

File: test/python/test_qasm_parser.py
Comment (3yakuya) [reply test/python/test_qasm_parser.py:89]: Alright, got it, this is a useful lesson overall, thank you!

File: test/python/qasm/example_version_2.qasm
Comment (jakelishman) [reply test/python/qasm/example_version_2.qasm:1]: That's a good idea, let's leave it as it is.

File: test/python/qasm/example_version_fail.qasm
Comment (jakelishman) [reply test/python/qasm/example_version_fail.qasm]: Don't worry about it - it's not really true anyway. Before you made the PR, I didn't know how hard or easy it would be to fix.  After you'd made it, all I had to do was check that what you'd done was sensible, not actually go through and do it myself, and you did this well.

Finding extra special test cases, catching weirdnesses with `re` and things like that is just more experience, and it's why we have review anyway.

**Review by jakelishman — APPROVED** [review]
I pushed up a couple of commits just to simplify the creation of some strings, and fix a typo in the test case (which was probably my fault anyway).  This is good to merge now, and it's stable for backport, so we'll put it out in Terra 0.19.2 whenever we get around to releasing that :tada:

File: test/python/qasm/example_version_fail.qasm
Comment (3yakuya) [reply test/python/qasm/example_version_fail.qasm]: Well I appreciate the encouragement and would continue to work on this anyways :D

**Pull request merged**
